File: scm-java/src/main/scheme/primitives/PrimitiveAdd.java
package scheme.primitives;

import scheme.*;

public class PrimitiveAdd extends Primitive {
    @Override
    public String name() {
        return "+";
    }

    @Override
    public String info() {
        return "Syntax: (+ z1 ...)\n" +
               "Library: (scheme base)\n" +
               "Description: Returns the sum of its arguments. With no arguments, returns 0.\n" +
               "Example:\n" +
               "  (+ 3 4) => 7\n" +
               "  (+ 3) => 3\n" +
               "  (+) => 0";
    }

    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        if (arguments.length == 0) return 0L;
        if (allIntegers(arguments)) {
            Object result = arguments[0];
            for (int i = 1; i < arguments.length; i++) {
                result = IntegerMath.genericAdd(result, arguments[i]);
            }
            return result;
        } else if (allExactNums(arguments)) {
            Object result = arguments[0];
            for (int i = 1; i < arguments.length; i++)
                result = Rational.add(result, arguments[i]);
            return result;
        } else if (hasComplex(arguments)) {
            Object result = arguments[0];
            for (int i = 1; i < arguments.length; i++)
                result = Complex.add(result, arguments[i]);
            return result;
        } else {
            double result = toReal(arguments[0]);
            for (int i = 1; i < arguments.length; i++) {
                result += toReal(arguments[i]);
            }
            return result;
        }
    }
}

File: scm-csharp/scheme/primitives/PrimitiveAdd.cs
namespace scheme;

public class PrimitiveAdd : Primitive
{
    public override string Name()
    {
        return "+";
    }

    public override string Info()
    {
        return
            "Syntax: (+ z1 ...)\n" +
            "Library: (scheme base)\n" +
            "Description: Returns the sum of its arguments. With no arguments, returns 0.\n" +
            "Example:\n" +
            "  (+ 3 4) => 7\n" +
            "  (+ 3) => 3\n" +
            "  (+) => 0";
    }
    
    public override object Apply(SourcePos? pos, object[] arguments)
    {
        if (arguments.Length == 0) return 0L;
        if (AllIntegers(arguments))
        {
            object result = arguments[0];
            for (int i = 1; i < arguments.Length; i++)
            {
                result = IntegerMath.GenericAdd(result, arguments[i]);
            }
            return result;
        }
        else if (AllExactNums(arguments))
        {
            object result = arguments[0];
            for (int i = 1; i < arguments.Length; i++)
                result = Rational.Add(result, arguments[i]);
            return result;
        }
        else if (HasComplex(arguments))
        {
            object result = arguments[0];
            for (int i = 1; i < arguments.Length; i++)
                result = Complex.Add(result, arguments[i]);
            return result;
        }
        else
        {
            double result = ToReal(arguments[0]);
            for (int i = 1; i < arguments.Length; i++)
            {
                result += ToReal(arguments[i]);
            }
            return result;
        }
    }
}
